substrate/client/network/sync/src/strategy/state.rs

```diff
@@ -0,0 +1,758 @@
+		is_best.then_some({
+			let best_number = *announce.header.number();
+			let best_hash = announce.header.hash();
+			if let Some(ref mut peer) = self.peers.get_mut(&peer_id) {
+				peer.best_number = best_number;
+			}
+			// Let `SyncingEngine` know that we should update the peer info.
+			(best_hash, best_number)
+		})
+	}
+
+	/// Process state response.
+	pub fn on_state_response(&mut self, peer_id: PeerId, response: OpaqueStateResponse) {
+		if let Err(bad_peer) = self.on_state_response_inner(peer_id, response) {
+			self.actions.push(StateStrategyAction::DropPeer(bad_peer));
+		}
+	}
+
+	fn on_state_response_inner(
+		&mut self,
+		peer_id: PeerId,
+		response: OpaqueStateResponse,
+	) -> Result<(), BadPeer> {
+		if let Some(peer) = self.peers.get_mut(&peer_id) {
+			peer.state = PeerState::Available;
+		}
+
+		let response: Box<StateResponse> = response.0.downcast().map_err(|_error| {
+			error!(
+				target: LOG_TARGET,
+				"Failed to downcast opaque state response, this is an implementation bug."
+			);
+
+			BadPeer(peer_id, rep::BAD_RESPONSE)
+		})?;
+
+		debug!(
+			target: LOG_TARGET,
+			"Importing state data from {} with {} keys, {} proof nodes.",
+			peer_id,
+			response.entries.len(),
+			response.proof.len(),
+		);
+
+		match self.state_sync.import(*response) {
+			ImportResult::Import(hash, header, state, body, justifications) => {
+				let origin = BlockOrigin::NetworkInitialSync;
+				let block = IncomingBlock {
+					hash,
+					header: Some(header),
+					body,
+					indexed_body: None,
+					justifications,
+					origin: None,
+					allow_missing_state: true,
+					import_existing: true,
+					skip_execution: true,
+					state: Some(state),
+				};
+				debug!(target: LOG_TARGET, "State download is complete. Import is queued");
+				self.actions
+					.push(StateStrategyAction::ImportBlocks { origin, blocks: vec![block] });
+				Ok(())
+			},
+			ImportResult::Continue => Ok(()),
+			ImportResult::BadResponse => {
+				debug!(target: LOG_TARGET, "Bad state data received from {peer_id}");
+				Err(BadPeer(peer_id, rep::BAD_STATE))
+			},
+		}
+	}
+
+	/// A batch of blocks have been processed, with or without errors.
+	///
+	/// Normally this should be called when target block with state is imported.
+	pub fn on_blocks_processed(
+		&mut self,
+		imported: usize,
+		count: usize,
+		results: Vec<(Result<BlockImportStatus<NumberFor<B>>, BlockImportError>, B::Hash)>,
+	) {
+		trace!(target: LOG_TARGET, "State sync: imported {imported} of {count}.");
+
+		let results = results
+			.into_iter()
+			.filter_map(|(result, hash)| {
+				if hash == self.state_sync.target_hash() {
+					Some(result)
+				} else {
+					debug!(
+						target: LOG_TARGET,
+						"Unexpected block processed: {hash} with result {result:?}.",
+					);
+					None
+				}
+			})
+			.collect::<Vec<_>>();
+
+		if !results.is_empty() {
+			// We processed the target block
+			results.iter().filter_map(|result| result.as_ref().err()).for_each(|e| {
```

Comment:
I'm not sure that this can be done. We also need to finish the strategy if the target block import has failed, but if we filter out unsuccessful results we won't be able to distinguish between the following situations:
1. No target block was processed
2. Target block processing has failed

Am I missing something?

Reply:
Yeah, my mistake, but we could `filter_map` first loop into bool, and do logging on the go in the single pass. Again it is minor :),  feel free to ignore this.